// sonos-discovery/lib/helpers/logger.js
'use strict';

var util = require('util');

var LOG_LEVEL = Object.freeze({
  TRACE: 10,
  DEBUG: 20,
  INFO: 30,
  WARN: 40,
  ERROR: 50
});

function isValidLevel(level) {
  var configuredLogLevel = process.env.NODE_LOG_LEVEL || 'info';
  return LOG_LEVEL[level] >= LOG_LEVEL[configuredLogLevel.toUpperCase()];
}

function log(level) {
  if (!isValidLevel(level)) {
    return;
  }

  var args = Array.from(arguments).slice(1);
  var errors = args.filter(function (x) {
    return x instanceof Error;
  });
  var objects = args.filter(function (x) {
    return !(x instanceof Error) && x instanceof Object;
  });
  var message = args.filter(function (x) {
    return !(x instanceof Object);
  }).join(' ');
  var output = [new Date().toISOString(), level, message];

  console.log(output.join(' '));
  errors.forEach(function (x) {
    return console.log('  ' + x.stack);
  });
  objects.forEach(function (x) {
    return console.log('  ' + util.inspect(x));
  });
}

var logger = {
  trace: log.bind(null, 'TRACE'),
  debug: log.bind(null, 'DEBUG'),
  info: log.bind(null, 'INFO'),
  warn: log.bind(null, 'WARN'),
  error: log.bind(null, 'ERROR')
};

module.exports = logger;
