// sonos-discovery/lib/services/soundcloud.js
'use strict';

var request = require('request-promise');
var clientId = '6b9ec970f07f410376f1db1dfa8d71b3';
var apiBaseEndpoint = 'http://api.soundcloud.com';
var settings = {};

try {
  settings = require.main.require('./settings.json');
} catch (e) {}

settings.soundcloud = settings.soundcloud || clientId;

var endpoints = {
  track: '/tracks/'
};

function parseUri(uri) {
  // x-sonos-http:track%3a232202756.mp3?sid=160&flags=8224&sn=10
  var id = void 0;
  var uriParts = /x-sonos-http:track%3a(\d+)\./.exec(uri);
  if (uriParts) {
    id = uriParts[1];
  }

  return {
    id: id
  };
}

function tryGetHighResArt(uri) {
  var trackInfo = parseUri(uri);

  var apiUrl = [apiBaseEndpoint, endpoints.track, trackInfo.id, '?client_id=', settings.soundcloud].join('');

  return request({
    url: apiUrl,
    json: true
  }).then(function (response) {
    var artwork = void 0;

    // jscs:disable requireCamelCaseOrUpperCaseIdentifiers
    /* jshint -W106 */
    if (response.artwork_url) {
      artwork = response.artwork_url.replace('large', 't500x500');
    }

    // jscs:enable requireCamelCaseOrUpperCaseIdentifiers
    /* jshint +W106 */

    return artwork;
  });
}

module.exports = function (services) {
  services[160] = {
    tryGetHighResArt: tryGetHighResArt
  };
};
